lib/ansible/module_utils/basic.py

```diff
@@ -1832,9 +1834,19 @@ def safe_eval(self, value, locals=None, include_exceptions=False):
     def _check_type_str(self, value):
         if isinstance(value, string_types):
             return value
-        # Note: This could throw a unicode error if value's __str__() method
-        # returns non-ascii.  Have to port utils.to_bytes() if that happens
-        return str(value)
+
+        # Ignore, warn, or error when converting to a string.
+        # The current default is to warn. Change this in Anisble 2.12 to error.
+        common_msg = 'The value {0!r} (type {0.__class__.__name__}) in a string field was converted to {1!r} (type string). {2}'
+        if self._string_conversion_action == 'error':
+            extra_msg = 'Quote the entire value to ensure it does not change.'
+            msg = common_msg.format(value, to_text(value), extra_msg)
+            self.fail_json(msg=msg)
```

Comment:
Why not raise a `TypeError` (like the other `_check_type_*` functions)? Then `_check_argument_types` prepends `argument %s is of type %s and we were unable to convert to %s:` to the message.

Reply:
Good point. I will do that to be consistent with other methods.